module/state_synchronization/requester/status/status.go

```diff
@@ -0,0 +1,192 @@
+package status
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/storage"
+	"github.com/rs/zerolog"
+)
+
+// Status encapsulate the ExecutionDataRequester's state and provides methods to consistently
+// manage the requester's responsibilities.
+type Status struct {
+	// The highest block height for which notifications have been sent
+	lastNotified uint64
```

Comment:
To me, having this stateful `Status` module is  extra complexity and not crash-resistant. A better way is to go stateless.

We could actually go stateless by creating another jobqueue, and persists the notified height. This is how I think it could work:

1. The jobqueue listens to `OnBlockFinalized` event to check if there is new height being sealed. If there is, then create worker to request its ExecutionData.
2. If a worker finished downloading, it marks this height has been done in the database, and notify the jobqueue for the job completion. And the jobqueue will check if this height is the next unprocessed height, if is then update the `processed index` in the database, and create worker to work on next job. 
3. When a worker starts, it first check if the ExecutionData is already stored in the database, if yes, then report job completion right away. This allows the processed index to go fast forward when a job to fill the gap in the unprocessed height is done. 
4. When EN is crashed or interrupted, the jobqueue will restart from the next unprocessed height, and create worker to work on it again, this ensures data integrity that execution data for every height will eventually be downloaded.
5. We also configure the jobqueue, so that it stops creating more workers if there is a gap in height is never filled. 
6. Now the problem is that we have the constraint that the worker (requester) should notify some consumer about execution data downloads in block order. I think this is where the stateful Status module is still being used, which I think could go stateless. To go stateless, we could create another jobqueue (NotifyJobQueue) for notifying downloaded height, it only stores a single height in the database, whenever the request worker finishes, it notify the NotifyJobQueue. It's the same way we notify the RequestJobQueue when a new height is sealed. So the NotifyJobQueue will create a single worker, which read the RequestJobQueue's processed height, if it's higher than notified height, then we read the ExecutionData from storage for the next un-notified height, and forward it to the event consumer. We can guarantee the event consumer receives the ExecutionData in block order. 
7. Since it's stateless, another benefit is that it's crash-resistant, on startup, the NotifyJobQueue will check if there un-notified height to notify, which ensures each height will eventually be notified in block order. Whereas, if we store it in Status module which is stateful, it's not crash-resistant.

Reply:
updated in eaf25f3